stack/stack.go

```diff
@@ -26,7 +26,7 @@ func (s *S) Init(size int) {
 func (s *S) Push(v interface{}) {
 	// dynamically increase the size of storage as needed
 	if s.i+1 == cap(s.storage) {
-		ns := make([]interface{}, s.i*2)
+		ns := make([]interface{}, cap(s.storage)*2)
```

Comment:
No sure if I got the comment. But the fix is for the case when the stack is init with 1, in that case s.i is 0 so we end up creating a slice with 0 cap and the Push fails (see unit test added). So we need to multiple by the cap or we can multiple by (s.i + 1) but IMO that is little less readable.